staging/cse/windows/nvidiagpudriverfunc.ps1

```diff
@@ -0,0 +1,169 @@
+        [Parameter(Mandatory = $false)]
+        [string]$GpuDriverURL
+    )
+  
+    if (-not $EnableInstall) {
+        Write-Log "ConfigGPUDriverIfNeeded is false. GPU driver installation skipped as per configuration."
+        return
+    }
+  
+    Write-Log "ConfigGPUDriverIfNeeded is true. GPU driver installation started as per configuration."
+  
+    if ([string]::IsNullOrEmpty($GpuDriverURL)) {
+        $ErrorMsg = "DriverURL is not properly specified."
+        Write-Log $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_URL_NOT_SET -ErrorMessage $ErrorMsg
+        return
+    }
+
+    $fileName = [IO.Path]::GetFileName($GpuDriverURL)
+    if (-not $fileName.EndsWith(".exe")) {
+        $ErrorMsg = "DriverURL does not point to a exe file"
+        Write-Log $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_URL_NOT_EXE -ErrorMessage $ErrorMsg
+        return
+    }
+
+    $Target = "C:\AzureData\$fileName"
+    $LogFolder = "C:\AzureData"
+
+    Write-Log "Attempting to install Nvidia driver..."
+
+    Write-Log "Downloading from $GpuDriverURL to $Target"
+    DownloadFileOverHttp -Url $GpuDriverURL -DestinationPath $Target -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_DOWNLOAD_FAILURE
+    Write-Log "Installer download complete"
+    
+    $IsSignatureValid = VerifySignature $Target 
+    if ($IsSignatureValid -eq $false) {
+        $ErrorMsg = "Signature embedded in $($Target) is not valid."
+        Write-Log $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_INVALID_SIGNATURE -ErrorMessage $ErrorMsg
+        return
+    }
+    else {
+        Write-Log "Signature embedded in $($Target) is valid."
+    }
+
+    Write-Log "Installing $Target ..."
+    try {
+        $InstallLogFolder = "$LogFolder\NvidiaInstallLog"
+        $Arguments = "-s -n -log:$InstallLogFolder -loglevel:6"
+    
+        $p = Start-Process -FilePath $Target -ArgumentList $Arguments -PassThru
+        
+        $Timeout = 10 * 60 # in seconds. 10 minutes for timeout of the installation
+
+        # This is for testability. Start-Process mock returns a hashtable.
+        if (-not ($p -is [hashtable])) {
+            Wait-Process -InputObject $p -Timeout $Timeout -ErrorAction Stop
+        }
+    
+        # check if installation was successful
+        if ($p.ExitCode -eq 0 -or $p.ExitCode -eq 1) {
+            # 1 is issued when reboot is required after success
+            Write-Log "GPU Driver Installation Success. Code: $($p.ExitCode)"
+            Remove-InstallerFile -InstallerPath $Target
+        } 
+        else {
+            $ErrorMsg = "GPU Driver Installation Failed! Code: $($p.ExitCode)"
+            Write-Log $ErrorMsg
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+            return
+        }
+
+        # check if reboot is needed
```

Comment:
Suggest to pass $p.ExitCode to the function Handle-InstallationResult and only set $global:RebootNeeded inside Handle-InstallationResult .

The logical should be more clear.
Fail if $gpuInstallationExitCode is neither 0 nor 1
Reboot if $p.ExitCode is 1
Reeboot if $p.ExitCode is 0 and vmsize is  NVseries